[chore][exporter/elasticsearch] Reduce heap allocs in profiles parsing

rockdaboot opened the pull request:
#### Description
Go frames do not have the `process.executable.build_id.htlhash` attribute.
This means that via `getBuildID()` -> `getStringFromAttribute()`, for every Go frame a complex error message is created (heap allocations involved) and then discarded.

This PR returns a constant error message ("missing attribute") to avoid these heap allocations. It is left to the caller of `getStringFromAttribute()` to add additional information, if required.

rockdaboot committed:
[chore][exporter/elasticsearch] Reduce heap allocs in profiles parsing

florianl commented on exporter/elasticsearchexporter/internal/serializer/otelserializer/serializeprofiles/transform.go:
The change clames to handle missing `process.executable.build_id.htlhash` attributes:
```
Go frames do not have the process.executable.build_id.htlhash attribute.
```

But the proposed change is changing the error returned for the attribute `profile.frame.type`.

rockdaboot commented on exporter/elasticsearchexporter/internal/serializer/otelserializer/serializeprofiles/transform.go:
```diff
@@ -384,8 +384,7 @@ func getStringFromAttribute(dic pprofile.ProfilesDictionary, record attributable
 		}
 	}
 
-	return "", fmt.Errorf("failed to get '%s' from indices %v: %w",
```
As said in the description of th PR, the caller should be responsible to log this kind of information, as it has all the information. In other words, the error message doesn't add any value, as `attrKey` and `record` are known to the caller.
The only additional value is `errMissingAttribute`.

Especially, when you look at the two callers:
1. `getBuildID()`: It discards the error message completely. Building the error message is "heavy lifting" and includes heap allocations - this is completely superfluous. And it is done for every single Go frame, why???
2. `stackFrames()`: Here, the `attrKey` is always `profile.frame.type`. This is valuable information and added to the error return (L295, see above). I am not sure how adding the list of attribute indices help with debugging here, if you don't know the attributeTable? We already know that the searched-for `attrKey` isn't reflected in the indices.

If you still think the indices are valuable for debugging, we can change the error message for point 2. Because this error isn't triggered under normal circumstances, it wouldn't hurt (but still, how useful would this ever be?).

florianl commented:
As my [comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/41555#discussion_r2245700926) gets hidden, I'm posting here:

Reading your argument and following code paths, I think, the issue around errors returned by `getStringFromAttribute()` is just the symptom, that `getBuildID()` is called for **every** frame. I focused too much an Go handling, that I ignored, `getBuildID()` is called for every frame - regardless if the frame i native or some kind of interpreted frame (where Go would be counted as interpreted frame).
For any kind of interpreted frame, `getBuildID()` will fail. And this sums up and maybe these are the heap allocations you noticed.

So, I think, `getFrameID()` should be `frameType` aware and not call `getBuildID()` at all and directly synthesize an ID for interpreted frame types.

rockdaboot commented:
> As my [comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/41555#discussion_r2245700926) gets hidden, I'm posting here:
> 
> Reading your argument and following code paths, I think, the issue around errors returned by `getStringFromAttribute()` is just the symptom, that `getBuildID()` is called for **every** frame. I focused too much an Go handling, that I ignored, `getBuildID()` is called for every frame - regardless if the frame i native or some kind of interpreted frame (where Go would be counted as interpreted frame). For any kind of interpreted frame, `getBuildID()` will fail. And this sums up and maybe these are the heap allocations you noticed.
> 
> So, I think, `getFrameID()` should be `frameType` aware and not call `getBuildID()` at all and directly synthesize an ID for interpreted frame types.

I get your point.
My thoughts are: we can't forecast the changes in the agent (could be ebpf-profiler or something else; also any version of agent) and this piece of code should be as flexible as possible. Special casing by frameType will require changes here in sync with changes to the agent - and needs to keep backwards and forewards compatibility (there is no version check currently). We know that server side and client are often updated at different times in larger deployments.
All this isn't a problem when checking the buildID; the current code does it correctly: generating the buildID if it isn't available. It's obviously a better, more resilient approach than working with frameType.

florianl commented:
While the special handling based on frame type wasn't ideal, our discussion revealed a broader, specification-violating issue with this collector, regardless of its source.

The attribute `process.executable.build_id.htlhash` should only be attached to [mappings](https://opentelemetry.io/docs/specs/otel/profiles/mappings/). And mappings are optional to locations in the [current](process.executable.build_id.htlhash) and [upcoming](https://github.com/open-telemetry/opentelemetry-proto/blob/6deb067376dc017240f940c252b8b116fd3becf1/opentelemetry/proto/profiles/v1development/profiles.proto#L452-L456) version of the protocol.

So, calling `getBuildID` will fail in most cases at this point:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b4fbd61c38a24fba0fa52b70dc271173302f36eb/exporter/elasticsearchexporter/internal/serializer/otelserializer/serializeprofiles/transform.go#L328-L331

> Another point that speaks against logging the unbounded array of indices

I agree with that.
An alternative could be `return "", fmt.Errorf("failed to get '%s': %w," attrKey, errMissingAttribute)`.
This avoid unbounded and extensive logging but still provides necessary information. It also removes the burden to the call of the function `getBuildID()` to add costum logging around the not found attribute key.

rockdaboot committed:
Merge remote-tracking branch 'upstream/main' into esex-reduce-allocs

rockdaboot committed:
Only call getBuildID() if location has a mapping

rockdaboot commented:
> The attribute process.executable.build_id.htlhash should only be attached to [mappings](https://opentelemetry.io/docs/specs/otel/profiles/mappings/). And mappings are optional to locations in the [current](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/process.executable.build_id.htlhash) and [upcoming](https://github.com/open-telemetry/opentelemetry-proto/blob/6deb067376dc017240f940c252b8b116fd3becf1/opentelemetry/proto/profiles/v1development/profiles.proto#L452-L456) version of the protocol.

That's a good point, addressed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/41555/commits/aaaea85eb48a9dd2bc0e3c1e7744201d14fde8b9.

> An alternative could be return "", fmt.Errorf("failed to get '%s': %w," attrKey, errMissingAttribute).

Added. Though in general, I'd prefer if the caller can decide whether a heap allocation on error is preferred or not. But this isn't really relevant any more with the changes in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/41555/commits/aaaea85eb48a9dd2bc0e3c1e7744201d14fde8b9.

rockdaboot commented:
> Is this ready to merge?

@atoulme Yes, it ready to merge
